internal/provider/resource_host_catalog_plugin.go

```diff
@@ -0,0 +1,413 @@
+package provider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/boundary/api"
+	"github.com/hashicorp/boundary/api/hostcatalogs"
+	"github.com/hashicorp/go-secure-stdlib/parseutil"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+const (
+	hostCatalogTypePlugin = "plugin"
+)
+
+func resourceHostCatalogPlugin() *schema.Resource {
+	return &schema.Resource{
+		Description: "The host catalog resource allows you to configure a Boundary plugin-type host catalog. Host " +
+			"catalogs are always part of a project, so a project resource should be used inline or you " +
+			"should have the project ID in hand to successfully configure a host catalog.",
+
+		CreateContext: resourceHostCatalogPluginCreate,
+		ReadContext:   resourceHostCatalogPluginRead,
+		UpdateContext: resourceHostCatalogPluginUpdate,
+		DeleteContext: resourceHostCatalogPluginDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			IDKey: {
+				Description: "The ID of the host catalog.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			NameKey: {
+				Description: "The host catalog name. Defaults to the resource name.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			DescriptionKey: {
+				Description: "The host catalog description.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			ScopeIdKey: {
+				Description: "The scope ID in which the resource is created.",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			PluginIdKey: {
+				Description: "The ID of the plugin that should back the resource. This or " + PluginNameKey + " must be defined.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Computed:    true, // If name is provided this will be computed
+			},
+			PluginNameKey: {
+				Description: "The name of the plugin that should back the resource. This or " + PluginIdKey + " must be defined.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Computed:    true,
+			},
+			TypeKey: {
+				Description: "The host catalog type. Only `plugin` is supported, and is the default.",
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Optional:    true,
+				Default:     "plugin",
+			},
+			AttributesJsonKey: {
+				Description: `The attributes for the host catalog. Either values encoded with the "jsonencode" function, pre-escaped JSON string, or a file:// or env:// path. Set to a string "null" or remove the block to clear all attributes in the host catalog.`,
+				Type:        schema.TypeString,
```

Comment:
Why not use `schema.TypeMap`?   I think we'd want the attributes defined directly in TF (unlike secrets which might be better to be passed in through a envar or path) and from a very quick look at the TypeMap it seems like it would be some of the diff checks for you.

Reply:
That is a very long conversation, more than I want to have here (so Slack or Zoom) but basically: TypeMap, despite being based on HCL and despite coming in as a `map[string]interface{}`, is not actually valid for anything other than strings. So if what you're providing isn't something that fits into `map[string]string` the behavior is more or less undefined or just flat out doesn't work.